### ISSUE TITLE
task1_ai_ml_model : Initial Research on Rasa as tool for NLU development in the context of the project

### ISSUE THREAD

**Issue by nlauchande** [omdena/ptsd#1]
*Goals of the task* : 

Research on : https://rasa.com/docs/

1. Execute a basic getting started tutorial

2. Understand the architecture of a conversational NLP solution 

3. Analyse tradeoffs  w/ Dialogflow

4. Sketch a possible solution for a simple conversational bot

**Comment by LaurenSpiegel** [comment]
**_TLDR_**: Lots of built-in tools to get us going quickly and flexibility to switch in our own models.

**License**:

- Main Rasa software is under Apache 2.0 license. 
- Rasa X has a specially drafted license. Basically, can use Rasa X for non-competing purposes. If bring patent claim against Rasa Technologies, lose all rights.

**Docs**:

- Links in published docs are messed up. Use github repo directly: https://github.com/RasaHQ/rasa/tree/master/docs/user-guide 
- Or this link works (using sidebar to navigate):
https://rasa.com/docs/rasa/

**Integrations**:
- Off the shelf integrations with Slack, FB messenger, Telegram, Twilio, etc. -- https://rasa.com/docs/rasa/user-guide/messaging-and-voice-channels/

**Structure**:

- Overall architecture diagram here -- https://rasa.com/docs/rasa/user-guide/architecture/
- Has **“NLU models”** -- turns messages into structured data
  - Performs intent classification and entity extraction
  - "I am looking for a Mexican restaurant in the center of town" becomes:
`{
  "intent": "search_restaurant",
  "entities": {
    "cuisine" : "Mexican",
    "location" : "center"
  }
}`
  - Training data can be in markdown or json.
  - Looks difficult to assemble training data.
  - Instead of creating synthetic examples they recommend using lookup tables for synonyms.
  - The NLU model is really a pipeline of components. 
     - There are pre-made pipelines and can construct your own.
     - pretrained_embeddings_spacy pipeline uses pre-trained word vectors from either GloVe or fastText
     - supervised embeddings pipeline doesn’t use any pre-trained word vectors, but instead fits these specifically for your dataset
    - Lots of options for components -- https://rasa.com/docs/rasa/nlu/components/
    - Can make custom components (like a sentiment classifier) -- https://rasa.com/docs/rasa/api/custom-nlu-components/#custom-nlu-components
    - Tutorial on building own component: https://blog.rasa.com/enhancing-rasa-nlu-with-custom-components/
 - Has **“Core models”** -- handles how bot responds (“dialogue management”) 
   - Training data are “stories” -- example dialogues.
   - Can import stories or use ui to create training data.
   - The “Policy” class decides what action to take at each step in the conversation. If have multiple policies, will use prediction with the highest confidence.
     - There is a keras policy with an LSTM as the default architecture. You can override with your own architecture
     - There is an embedding policy which they created and published a paper on--https://rasa.com/docs/rasa/core/policies/#embedding-policy
     - Memoization Policy has a 1.0 confidence if same conversation exists in training data. 
     - There is a form Policy which is an extension of this which could be helpful for us.
     - Mapping Policy maps an intent to an action.
     - Fallback policy invokes a fallback action if the intent recognition has confidence below a given threshold.
 - For debugging and training there is an interactive learning mode where you interact with bot and give feedback on responses.  
- There is also a visualization of conversations tool which is helpful. https://rasa.com/docs/rasa/core/interactive-learning/


**Bonuses**:
- Includes scripts to NLU and Core models.
- Can run anywhere and has docker instructions to package up.
- For tracker store can use mongo, redis, postgresql or custom -- https://rasa.com/docs/rasa/user-guide/running-rasa-with-docker/
- There is support for storing models no s3, GCP or Azure -- https://rasa.com/docs/rasa/user-guide/cloud-storage/
- Can migrate from dialogflow to rasa -- https://rasa.com/docs/rasa/migrate-from/google-dialogflow-to-rasa/
  - They claim people switch because:
    - Faster since can run locally
    - Can tune models to get higher accuracy
    - No vendor lock-in
      - Even though you can migrate from dialogflow, you are really just moving your data out. So the model will still be in dialogflow and your performance could be much different once in rasa.
- Active community with forum -- https://forum.rasa.com/

**Comment by elateifsara** [comment]
Thank you @LaurenSpiegel extremely helpful.

**Comment by SubmitCode** [comment]
@LaurenSpiegel thank you for the great summary.

**Comment by MamadouBousso** [comment]
A draft of an md file for RASA. I put it in a docx file because the extension is not supported
[intentPSTD.docx](https://github.com/omdena/ptsd/files/3390469/intentPSTD.docx)
